ush/python/pygw/src/pygw/schema.py

```diff
@@ -0,0 +1,891 @@
+            and self._schema == other._schema
+        )
+
+    def reset(self):
+        if hasattr(self._schema, "reset"):
+            self._schema.reset()
+
+
+class Hook(Schema):
+    def __init__(self, *args, **kwargs):
+        self.handler = kwargs.pop("handler", lambda *args: None)
+        super(Hook, self).__init__(*args, **kwargs)
+        self.key = self._schema
+
+
+class Forbidden(Hook):
+    def __init__(self, *args, **kwargs):
+        kwargs["handler"] = self._default_function
+        super(Forbidden, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def _default_function(nkey, data, error):
+        raise SchemaForbiddenKeyError(
+            "Forbidden key encountered: %r in %r" % (nkey, data), error)
+
+
+class Literal(object):
+    def __init__(self, value, description=None):
+        self._schema = value
+        self._description = description
+
+    def __str__(self):
+        return self._schema
+
+    def __repr__(self):
+        return 'Literal("' + self.schema + '", description="' + (self.description or "") + '")'
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def schema(self):
+        return self._schema
+
+
+class Const(Schema):
+    def validate(self, data, **kwargs):
+        super(Const, self).validate(data, **kwargs)
+        return data
+
+
+def _callable_str(callable_):
+    if hasattr(callable_, "__name__"):
+        return callable_.__name__
+    return str(callable_)
+
+
+def _plural_s(sized):
+    return "s" if len(sized) > 1 else ""
+
+# ----
+
+# The following functions are specific to global-workflow. The Schema
+# class module was obtained from:
+
+# https://github.com/keleshev/schema/blob/master/schema.py
+
+
+# TODO: Is there a better way to do this?
+logger = Logger("pygw.schema", colored_log=True)
+
+
+@logit(logger)
+def build_schema(yaml_path: str) -> Dict:
+    """
+    Description
+    -----------
+
+    This function parses a YAML-formatted file and defines the
+    respective schema.
+
+    Parameters
+    ----------
+
+    yaml_path: str
+
+        A Python string specifying the path to the YAML-formatted file
+        containing the schema attributes.
+
+    schema_dict: Dict
+
+        A Python dictionary containing the schema.
+
+    """
+
+    # Read the YAML-formatted file containing the schema.
+    data = yaml_file.parse_yaml(path=yaml_path)
+
+    schema_dict = {}
+    for datum in data:
+        data_dict = AttrDict(data[datum])
+
+        # Check whether the variable is optional; proceed accordingly.
+        if "optional" not in data_dict:
+            data_dict.optional = False
+        else:
+            pass
+
+        # Build the schema accordingly.
+        if data_dict.optional:
+            schema_dict[Optional(datum, data_dict.default)
+                        ] = locate(data_dict.type)
+
+        if not data_dict.optional:
+            schema_dict[datum] = locate(data_dict.type)
```

Comment:
Do you really need 2 ifs here?
Can you not use `if-else` or better still, `try-except`?
```suggestion
        if data_dict['optional']:
            schema_dict[Optional(datum, data_dict['default')] = locate(data_dict['type'])
        else:
            schema_dict[datum] = locate(data_dict['type'])
```

Reply:
I am not sure what the best path forward is here. The `if-else` makes the most sense to me. What are your thoughts on this? We could potentially add an infinite number of metadata options. If so, is `if-else` better than `try-except` or vice versa?

Reply:
lets go w/ `if-else`. If more come along, we can revisit.
I am curious about the use of `pydoc.locate` here.  Need to think why the function `type()` is not sufficient.